wellcomeml/ml/bert_classifier.py

```diff
@@ -1,236 +1,191 @@
+           validation_split(float): split for validation set during training, default 0.1
+           max_length(int): maximum number of tokens, controls padding and truncation, default 512
+           multilabel(bool): flag on whether the problem is multilabel i.e. Y a matrix or a column
+           random_seed(int): controls random seed for reproducibility
+           from_pt(bool): flag about pretrained model in pytorch or tensorflow
+        """
         self.learning_rate = learning_rate
-        self.n_iterations = n_iterations
+        self.epochs = epochs  # used to be n_iterations
+        self.batch_size = batch_size
         self.pretrained = pretrained
-        self.l2 = l2
+        self.threshold = threshold
         self.validation_split = validation_split
-
-    def _init_nlp(self):
-        if self.pretrained == "bert":
-            self.nlp = spacy.load("en_trf_bertbaseuncased_lg")
-        elif self.pretrained == "scibert":
-            name = "scibert-scivocab-uncased"
-            path = check_cache_and_download("scibert_scivocab_uncased")
-
-            nlp = TransformersLanguage(trf_name=name, meta={"lang": "en"})
-            nlp.add_pipe(nlp.create_pipe("sentencizer"))
-            nlp.add_pipe(TransformersWordPiecer.from_pretrained(nlp.vocab, path))
-            nlp.add_pipe(TransformersTok2Vec.from_pretrained(nlp.vocab, path))
-            self.nlp = nlp
+        self.max_length = max_length
+        self.multilabel = multilabel
+        self.random_seed = 42
+        self.from_pt = from_pt
+
+    def _init_model(self, num_labels=2):
+        config = {
+            "name": self.pretrained,
+            "from_pt": self.from_pt
+        }
+        if self.pretrained in PRETRAINED_CONFIG:
+            config = PRETRAINED_CONFIG[self.pretrained]
+
+        pretrained = config["name"]
+        from_pt = config["from_pt"]
+
+        self.tokenizer = BertTokenizer.from_pretrained(pretrained)
+        self.model = TFBertForSequenceClassification.from_pretrained(
+            pretrained, from_pt=from_pt, num_labels=num_labels)
+
+    def _transform_data(self, input_ids, attention_mask, labels=None):
+        input_data = {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask
+        }
+        if labels is not None:
+            return input_data, labels
         else:
-            logger.info(f"{self.pretrained} is not among bert, scibert")
-            raise
-        # TODO: Add a parameter for exclusive classes, non multilabel scenario
-        self.textcat = self.nlp.create_pipe(
-            "trf_textcat",
-            config={"exclusive_classes": False, "architecture": "sigmoid_last_hidden"},
-        )
-
-        self.nlp.add_pipe(self.textcat, last=True)
+            return input_data
 
-        for label in self.unique_labels:
-            self.textcat.add_label(label)
+    def _prepare_data(self, X, Y=None, shuffle=False, repeat=False):
+        X_vec = self.tokenizer.batch_encode_plus(
```

Comment:
If I am not mistaken, this was the bottleneck for my processing. If using in a CPU, it doesn't parallelise. I am not sure if it parallelise with multi-GPU either. Wonder if the new version of transformers/torch has fixed this.

Reply:
Good point, actually my next task is to investigate a bit the tokenisation in order to build a class that we can create a custom tokenizer. So maybe this will give us a way